File: dp/LC2681.py
# https://leetcode.cn/problems/power-of-heroes/
class Solution:
    def sumOfPower(self, nums: List[int]) -> int:
        MOD = 10**9+7
        nums.sort()
        n = len(nums)
        dp = 0  # sum of min(sequence ending with i)
        pf_dp = 0
        ans = 0
        for i in range(n):
            dp = nums[i] + pf_dp
            pf_dp = pf_dp + dp
            ans = (ans + nums[i] * nums[i] * dp) % MOD
        return ans

File: dp/LC2681.cpp
// https://leetcode.cn/problems/power-of-heroes/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int sumOfPower(vector<int>& nums) {
        const int MOD = 1e9+7;
        int ans = 0;
        int s = 0;
        sort(nums.begin(), nums.end());
        for(int num: nums) {
            s = (2ll*s + num) % MOD;
            ans = (1ll*num*s + ans) % MOD;
        }
        return ans;
    }
};